File: docs/languages/ruby/primes.rb
# primes.rb

# Returns true if n is prime, false otherwise.
# Uses trial division to find factors.
def is_prime?(n)
    if n < 2
        false
    elsif n == 2
        return true
    elsif n.even?
        return false
    else
        candidate = 3
        while candidate * candidate <= n
            return false if n % candidate == 0
            candidate += 2
        end
        return true
    end
end

# Returns the number of primes less than the integer n.
def primes_less_than(n)
    return (1..n-1).count {|i| is_prime?(i)}
end

# Which of the numbers from 0 to 30 are prime?
count = 0
30.times do |n| 
    if is_prime?(n)
        count += 1
        puts "#{n} <--- prime #{count}"
    else
        puts "#{n}"
    end
end

max = 10000 # 1229 for 1000
puts "Number of primes less than #{max}: #{primes_less_than(max)}"

File: docs/languages/go/primes.go
// primes.go

package main

import "fmt"

func main() {
    for i := 0; i <= 25; i++ {
        fmt.Println(i, is_prime(i))
    }

    fmt.Println(primes_less_than(10000)) // correct answer: 1229
}

// Returns true if the integer n is prime, and false otherwise.
func is_prime(n int) bool {
    if n < 2 {
        return false
    } else if n == 2 {
        return true
    } else if n % 2 == 0 {
        return false
    } else {
        candidate := 3
        for candidate * candidate <= n {
            if n % candidate == 0 {
                return false
            }
            candidate += 2
        }
        return true
    }
}

// Uses a named return parameter.
func primes_less_than(n int) (result int) {
    for i := 2; i < n; i++ {
        if is_prime(i) {
            result++
        }
    }
    return // required with named return parameter
}
